src/code/operation.js

```diff
@@ -242,14 +253,37 @@ export default class Operation {
         this.anchors = this.anchors.filter(function(listAnchor) {
             return listAnchor !== originalPortal.id;
         });
-        this.addAnchor(newPortal)
-        for (let link_ in this.links) {
+        this.addAnchor(newPortal);
+        let linksToRemove = [];
+        for (let link_ = 0; link_ < this.links.length; link_++) {
             if (this.links[link_].fromPortalId == originalPortal["id"]) {
-                this.links[link_].fromPortalId = newPortal["id"];
+                if (this.links[link_].toPortalId === newPortal["id"]) {
+                    console.log(`Operation: Removing link '${this.links[link_].ID}' while swapping because it would create a link with the same source and target.`);
+                    linksToRemove.push(this.links[link_]);
+                } else if (!this.containsLinkFromTo(newPortal["id"], this.links[link_].toPortalId)) {
+                    this.links[link_].fromPortalId = newPortal["id"];
+                } else {
+                    console.log(`Operation: Removing link '${this.links[link_].ID}' while swapping because it would duplicate an existing link in the operation.`);
+                    linksToRemove.push(this.links[link_]);
+                }
             } else if (this.links[link_].toPortalId == originalPortal["id"]) {
-                this.links[link_].toPortalId = newPortal["id"];
+                if (this.links[link_].fromPortalId === newPortal["id"]) {
+                    console.log(`Operation: Removing link '${this.links[link_].ID}' while swapping because it would create a link with the same source and target.`);
+                    linksToRemove.push(this.links[link_]);
+                } else if (!this.containsLinkFromTo(this.links[link_].fromPortalId, newPortal["id"])) {
+                    this.links[link_].toPortalId = newPortal["id"];
+                } else {
+                    console.log(`Operation: Removing link '${this.links[link_].ID}' while swapping because it would duplicate an existing link in the operation.`);
+                    linksToRemove.push(this.links[link_]);
+                }
             }
         }
+        // Remove the invalid links from the array (after we are done iterating through it)
+        this.links = this.links.filter(element => !linksToRemove.includes(element));
```

Comment:
Here you can use the binary operator `in` to test inclusion.
I like that you use filter here.